content/posts/cloud-services.md

```diff
@@ -0,0 +1,70 @@
+
+The virtual machine is totally unresponsive and the provider's metrics show a very high CPU load. He assigns a larger CPU type to the machine and restarts it. The shop is now responsive again, but still a bit slow. Kurt lets it be for the moment and starts to think about how this scenario can be avoided in the future. Back in the hotel he starts packing the application into a container and configures the cloud provider's **serverless container runtime**. Now the application automatically scales up when there's more demand. And even better, it automatically scales down to zero during the night, when nobody wants to buy theater tickets, which allows the theater group to save some money for other things.
+
+Serverless container runtimes and similar services let you use exactly as much resources as your applications need. This allows you not only to scale automatically, it also can safe you a lot of money if done right.
+
+Another advantage is that you don't have to manage a server anymore. Most of the time you just throw some code at it and the service knows what it needs to do.
+
+## Chapter 4: Queuing
+
+Before going on an ice fishing trip to Finland, Kurt wants to make sure he isn't going to be disturbed again. (And yes, there's [almost certainly high speed internet](https://www.telia.fi/asiakastuki/kuuluvuuskartta) on a Finish lake.) So he has a look at the shop's code and thinks of how it can be improved even more. Using the cloud provider's analysis tools he finds out that the app always immediately scales up when it creates the PDF ticket for an order. This also correlates with some complaints by customers that the ordering process usually takes pretty long.
+
+Kurt starts to draft a plan for refactoring the app. Instead of immediately producing the PDF and e-mail, the app now sends a message into the provider's **message queue** system. The code which produces the PDF and sends the e-mail now runs as a serverless function, triggered by the message. This means that the PDF production and e-mail sending is now run asynchronously in the background. The order process feels much faster now and the serverless container runs with a single instance most of the time.
+
+Using message queues you can vastly improve the performance and resource usage of an application. These messages can either be processed by your own application or you can define a serverless function which will be triggered by a message. Obviously this has a huge impact on the application's architecture and needs to be planned carefully.
+
+## Conclusion
+
+This was a really high level overview of some of the advantages a modern cloud provider can offer you. We learned about some of the major services all cloud providers offer in one way or another: virtual machines, object storage, managed databases, serverless runtimes and message queues.
```

Comment:
IMO this could be fleshed out a bit. I like the focus, but I believe there are more lessons to be learned. I.e.
- That you can constantly evolve your application to meet the current needs.
- That you actually can go on holidays and only need to take your laptop along, because you can access your infrastructure from anywhere there's internet.
- That you can offload operational work and thus achieve more and still go diving.

Reply:
Added:
> The major advantages that all services have in common are availability, scalability, adaptability, and security. These are all important points, which usually take quite some effort when you run them yourself.
Don't want to go deeper here. It would be another blog post. But it's still worth mentioning.